StartTime EndTime Flags don't wrap around - can block reboots

### Problem
If the startTime command flag is set to be after endTime kured can never reboot. timewindow.Contains() always returns false.

#### Potential Unit Test Showing the Problem 
You can add the following test case to `pkg/timewindow/timewindow_test.go` 
```go
	tests := []struct {
		days  string
		start string
		end   string
		loc   string
		cases []testcase
	}{
....
		{"mon,tue,wed,thu,fri", "9pm", "5am", "America/Los_Angeles", []testcase{
			{"2019/04/04 00:49 PDT", true}, // Would expect 12:49am to be between 9pm and 5am
			{"2019/04/04 22:49 PDT", true}, // Would expect 10:49pm to be between 9pm and 5am
			{"2019/04/05 08:59 PDT", false},
			{"2019/04/05 9:01 PDT", false},
			{"2019/03/31 10:00 PDT", false},
			{"2019/04/04 12:00 PDT", false},
		}},
```
And it will fail.

```bash
$ go test -timeout 30s github.com/weaveworks/kured/pkg/timewindow -run ^TestTimeWindows
--- FAIL: TestTimeWindows (0.00s)
    timewindow_test.go:64: (---MonTueWedThuFri--- between 21:00 and 05:00 America/Los_Angeles) contains (2019/04/04 00:49 PDT) didn't match expected result of true
    timewindow_test.go:64: (---MonTueWedThuFri--- between 21:00 and 05:00 America/Los_Angeles) contains (2019/04/04 22:49 PDT) didn't match expected result of true
FAIL
FAIL    github.com/weaveworks/kured/pkg/timewindow      0.006s
FAIL
```
If kured is setup this way nothing happens other timewindow.Contains() returns false for all times.

### Impact
A user of kured can improperly configure kured in way that will result in kured never rebooting for no obvious reason. Nothing is printed to the log. The startup of kured looks reasonable. (The following is for a slightly different configuration).

```bash
$ kubectl logs -n kube-system kured-hsgrx
time="2020-09-14T20:05:21Z" level=info msg="Kubernetes Reboot Daemon: 1.4.5"
time="2020-09-14T20:05:21Z" level=info msg="Node ID: <REDACTED>"
time="2020-09-14T20:05:21Z" level=info msg="Lock Annotation: kube-system/kured:weave.works/kured-node-lock"
time="2020-09-14T20:05:21Z" level=info msg="Reboot Sentinel: /var/run/reboot-required every 1h0m0s"
time="2020-09-14T20:05:21Z" level=info msg="Blocking Pod Selectors: []"
time="2020-09-14T20:05:21Z" level=info msg="Reboot on: SunMonTueWedThuFriSat between 21:00 and 05:00 America/Chicago"
time="2020-09-14T20:05:21Z" level=info msg="Force annotation cleanup after: 2h0m0s"
```

### Suggested Fix
A simple stop gap would be to include in the readme and other examples a warning that the startTime flag MUST be before the endTime flag. Wrap around is not possible.

Code fixes could include:
- Don't allow kured to start if startTime is after end time with a meaningful error message.
- Support time wrap around in timewindow.Contains.

--- a/pkg/timewindow/timewindow.go
+++ b/pkg/timewindow/timewindow.go
@@ -47,6 +47,19 @@ func (tw *TimeWindow) Contains(t time.Time) bool {
 	start := time.Date(loctime.Year(), loctime.Month(), loctime.Day(), tw.startTime.Hour(), tw.startTime.Minute(), tw.startTime.Second(), 0, tw.location)
 	end := time.Date(loctime.Year(), loctime.Month(), loctime.Day(), tw.endTime.Hour(), tw.endTime.Minute(), tw.endTime.Second(), 1e9-1, tw.location)
 
+	// Time Wrap validation
+	// First we check for start and end time, if start is after end time
+	// Next we need to validate if we want to wrap to the day before or to the day after
+	// For that we check the loctime value to see if it is before end time, we wrap with the day before
+	// Otherwise we wrap to the next day.
+	if tw.startTime.After(tw.endTime) {
+		if loctime.Before(end) {
+			start = start.Add(-24 * time.Hour)
+		} else {
+			end = end.Add(24 * time.Hour)
+		}
+	}
+
 	return (loctime.After(start) || loctime.Equal(start)) && (loctime.Before(end) || loctime.Equal(end))
 }
 

--- a/pkg/timewindow/timewindow_test.go
+++ b/pkg/timewindow/timewindow_test.go
@@ -20,12 +20,12 @@ func TestTimeWindows(t *testing.T) {
 		cases []testcase
 	}{
 		{"mon,tue,wed,thu,fri", "9am", "5pm", "America/Los_Angeles", []testcase{
-			{"2019/04/04 00:49 PDT", false},
-			{"2019/04/05 08:59 PDT", false},
-			{"2019/04/05 9:01 PDT", true},
 			{"2019/03/31 10:00 PDT", false},
+			{"2019/04/04 00:49 PDT", false},
 			{"2019/04/04 12:00 PDT", true},
 			{"2019/04/04 11:59 UTC", false},
+			{"2019/04/05 08:59 PDT", false},
+			{"2019/04/05 9:01 PDT", true},
 		}},
 		{"mon,we,fri", "10:01", "11:30am", "America/Los_Angeles", []testcase{
 			{"2019/04/05 10:30 PDT", true},
@@ -39,6 +39,43 @@ func TestTimeWindows(t *testing.T) {
 		{"mo,tu,we,th,fr", "00:00", "23:59:59", "UTC", []testcase{
 			{"2019/04/18 00:00 UTC", true},
 			{"2019/04/18 23:59 UTC", true},
+		}},
+		{"mon,tue,wed,thu,fri", "9pm", "5am", "America/Los_Angeles", []testcase{
+			{"2019/03/30 04:00 PDT", false},
+			{"2019/03/31 10:00 PDT", false},
+			{"2019/03/31 22:00 PDT", false},
+			{"2019/04/04 00:49 PDT", true},
+			{"2019/04/04 12:00 PDT", false},
+			{"2019/04/04 22:49 PDT", true},
+			{"2019/04/05 00:49 PDT", true},
+			{"2019/04/05 08:59 PDT", false},
+			{"2019/04/05 9:01 PDT", false},
+		}},
+		{"mon,tue,wed,thu,fri", "11:59pm", "00:01am", "America/Los_Angeles", []testcase{
+			{"2019/04/04 23:58 PDT", false},
+			{"2019/04/04 23:59 PDT", true},
+			{"2019/04/05 00:00 PDT", true},
+			{"2019/04/05 00:01 PDT", true},
+			{"2019/04/05 00:02 PDT", false},
+		}},
+		{"mon,tue,wed,fri", "11:59pm", "00:01am", "America/Los_Angeles", []testcase{
+			{"2019/04/04 23:58 PDT", false},
+			{"2019/04/04 23:59 PDT", false}, // Even that this falls in the between the hours Thursday is not included so should not run
+			{"2019/04/05 00:00 PDT", true},
+			{"2019/04/05 00:02 PDT", false},
+		}},
+		{"mon,tue,wed,thu", "11:59pm", "00:01am", "America/Los_Angeles", []testcase{
+			{"2019/04/04 23:58 PDT", false},
+			{"2019/04/04 23:59 PDT", true},
+			{"2019/04/05 00:00 PDT", false}, // Even that this falls in the between the hours Friday is not included so should not run
+			{"2019/04/05 00:02 PDT", false},
+		}},
+		{"mon,tue,wed,thu,fri", "11:59pm", "00:01am", "UTC", []testcase{
+			{"2019/04/04 23:58 UTC", false},
+			{"2019/04/04 23:59 UTC", true},
+			{"2019/04/05 00:00 UTC", true},
+			{"2019/04/05 00:01 UTC", true},
+			{"2019/04/05 00:02 UTC", false},
 		}},
 	}
 

